src/databricks/labs/ucx/aws/locations.py

```diff
@@ -0,0 +1,88 @@
+import logging
+from collections.abc import Iterable
+from pathlib import PurePath
+
+from databricks.labs.ucx.assessment.aws import AWSRoleAction
+from databricks.labs.ucx.aws.access import AWSResourcePermissions
+from databricks.labs.ucx.hive_metastore import ExternalLocations
+from databricks.labs.ucx.hive_metastore.locations import ExternalLocation
+
+from databricks.sdk import WorkspaceClient
+
+logger = logging.getLogger(__name__)
+
+
+class AWSExternalLocationsMigration:
+
+    def __init__(
+        self,
+        ws: WorkspaceClient,
+        external_locations: ExternalLocations,
+        aws_resource_permissions: AWSResourcePermissions,
+    ):
+        self._ws = ws
+        self._locations = external_locations
+        self._aws_resource_permissions = aws_resource_permissions
+
+    def run(self, location_init="UCX_location"):
+        # For each path find out the role that has access to it
+        # Find out the credential that is pointing to this path
+        # Create external location for the path using the credential identified
+        credential_dict = self._get_existing_credentials_dict()
+        external_locations = self._locations.snapshot()
+        existing_external_locations = self._ws.external_locations.list()
+        existing_paths = [external_location.url for external_location in existing_external_locations]
+        compatible_roles = self._aws_resource_permissions.load_uc_compatible_roles()
+        missing_paths = self._identify_missing_external_locations(external_locations, existing_paths, compatible_roles)
+        external_location_names = [external_location.name for external_location in existing_external_locations]
+        external_location_num = 1
+        for path, role_arn in missing_paths:
+            if role_arn not in credential_dict:
+                logger.error(f"Missing credential for role {role_arn} for path {path}")
+                continue
+            while True:
+                external_location_name = f"{location_init}_{external_location_num}"
+                if external_location_name not in external_location_names:
+                    break
+                external_location_num += 1
```

Comment:
can you turn these lines into a private method? there might be a list of infinite loop and it's always easier to find one with stack traces if it's in an isolated method